crates/krun/src/bin/krun.rs

```diff
@@ -294,90 +310,65 @@ fn main() -> Result<()> {
-            .expect("uid should not contain NUL character"),
-        CString::new(format!("{}", getgid().as_raw()))
-            .expect("gid should not contain NUL character"),
+    let mut krun_guest_args: Vec<String> = vec![
+        krun_guest_path,
+        username,
+        format!("{}", getuid().as_raw()),
+        format!("{}", getgid().as_raw()),
+        krun_server_path,
+        command
+            .to_str()
+            .context("Failed to process command as it contains invalid UTF-8")?
+            .to_string(),
     ];
-
-    krun_guest_args.push(krun_server_path);
-    krun_guest_args.push(
-        CString::new(
-            command
-                .to_str()
-                .context("Failed to process command as it contains invalid UTF-8")?,
-        )
-        .context("Failed to process command as it contains NUL character")?,
-    );
-    let command_argc = command_args.len();
     for arg in command_args {
-        let s = CString::new(arg)
-            .context("Failed to process command arg as it contains NUL character")?;
-        krun_guest_args.push(s);
+        krun_guest_args.push(arg);
     }
 
-    let krun_guest_args: Vec<*const c_char> = {
-        const KRUN_GUEST_ARGS_FIXED: usize = 4;
-        // SAFETY: All pointers must be stored in the same allocation.
-        // See https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html#safety
-        let mut vec = Vec::with_capacity(KRUN_GUEST_ARGS_FIXED + command_argc + 1);
-        for s in &krun_guest_args {
-            vec.push(s.as_ptr());
-        }
-        vec.push(std::ptr::null());
-        vec
-    };
-
     let mut env = prepare_env_vars(env).context("Failed to prepare environment variables")?;
     env.insert(
         "KRUN_SERVER_PORT".to_owned(),
         options.server_port.to_string(),
     );
-    let env: Vec<CString> = {
-        let mut vec = Vec::with_capacity(env.len());
-        for (key, value) in env {
-            let s = CString::new(format!("{key}={value}")).with_context(|| {
-                format!("Failed to process `{key}` env var as it contains NUL character")
-            })?;
-            vec.push(s);
-        }
-        vec
+
+    let mut krun_config = KrunConfig {
+        args: Vec::new(),
+        envs: Vec::new(),
     };
-    let env: Vec<*const c_char> = {
-        // SAFETY: All pointers must be stored in the same allocation.
-        // See https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html#safety
-        let mut vec = Vec::with_capacity(env.len() + 1);
-        for s in &env {
-            vec.push(s.as_ptr());
-        }
-        vec.push(std::ptr::null());
-        vec
+    for arg in krun_guest_args {
+        krun_config.args.push(arg);
+    }
+    for (key, value) in env {
+        krun_config.envs.push(format!("{}={}", key, value));
+    }
+    let krun_config = KrunBaseConfig {
+        config: krun_config,
     };
 
+    let mut config_file = NamedTempFile::new()
```

Comment:
Add a safety note here about the lifetime requirements of the `config_file` variable? The path will become invalid if this is dropped before krun is done with it.

Reply:
Done, thanks.